db/db_impl/db_impl.cc

```diff
@@ -1061,6 +1061,204 @@ FSDirectory* DBImpl::GetDataDir(ColumnFamilyData* cfd, size_t path_id) const {
+                                         ApplyReplicationLogRecordInfo* info) {
+  JobContext job_context(0, true);
+  Status s;
+
+  {
+    WriteThread::Writer w;
+    InstrumentedMutexLock l(&mutex_);
+    write_thread_.EnterUnbatched(&w, &mutex_);
+
+    switch (record.type) {
+      case ReplicationLogRecord::kMemtableWrite: {
+        mutex_.Unlock();
+        WriteBatch batch;
+        s = WriteBatchInternal::SetContents(&batch, std::move(record.contents));
+        assert(s.ok());
+
+        if (versions_->LastSequence() + 1 !=
+            WriteBatchInternal::Sequence(&batch)) {
+          std::ostringstream oss;
+          oss << "Gap in sequence numbers, expected="
+              << versions_->LastSequence() + 1
+              << " got=" << WriteBatchInternal::Sequence(&batch);
+          return Status::Corruption(oss.str());
+        }
+
+        SequenceNumber next_seq;
+        s = WriteBatchInternal::InsertInto(
+            &batch, column_family_memtables_.get(), &flush_scheduler_,
+            &trim_history_scheduler_,
+            true /* ignore_missing_column_families_ */, 0 /* log_number */,
+            this, false /* concurrent_memtable_writes */, &next_seq,
+            nullptr /* has_valid_writes */, seq_per_batch_, batch_per_txn_);
+        if (s.ok()) {
+          versions_->SetLastSequence(next_seq - 1);
+        }
+        mutex_.Lock();
+        break;
+      }
+      case ReplicationLogRecord::kMemtableSwitch: {
+        WriteContext write_context;
+        autovector<ColumnFamilyData*> cfds;
+        SelectColumnFamiliesForAtomicFlush(&cfds);
+        for (const auto cfd : cfds) {
+          cfd->Ref();
+          s = SwitchMemtable(cfd, &write_context, "");
+          cfd->UnrefAndTryDelete();
+          if (!s.ok()) {
+            break;
+          }
+        }
+        break;
+      }
+      case ReplicationLogRecord::kManifestWrite: {
+        Slice src = record.contents;
+        autovector<VersionEdit> edits;
+        DeserializeReplicationLogManifestWrite(&src, &edits);
+
+        auto mutable_options =
+            default_cf_handle_->cfd()->GetLatestMutableCFOptions();
+        std::unique_ptr<ColumnFamilyOptions> cf_options;
+
+        autovector<ColumnFamilyData*> cfds;
+        autovector<const MutableCFOptions*> mutable_cf_options_list;
+        autovector<autovector<VersionEdit*>> edit_lists;
+
+        std::vector<uint32_t> added_column_families;
+
+        auto current_update_sequence = versions_->GetManifestUpdateSequence();
+        uint64_t latest_applied_update_sequence = 0;
+        for (auto& e : edits) {
+          if (!e.HasManifestUpdateSequence()) {
+            return Status::InvalidArgument(
+                "Manifest write doesn't have a ManifestUpdateSequence");
+          }
+          latest_applied_update_sequence = e.GetManifestUpdateSequence();
+          if (e.GetManifestUpdateSequence() <= current_update_sequence) {
+            // Ignore the update, we already have it
```

Comment:
> This can happen since we always recover from the latest flushed kMemtableSwitch, so it's possible to apply same manifest update multiple times in the log.

Exactly. The internal design docs goes into some details on this, including some interesting graphics.

> Do we have a copy of the manifest file on local disk? Or for each manifest write, we will update the file in s3 directly?

Both is true. We have a local manifest copy and, in the case of durable replica, upload the manifest to the cloud on every Sync:
 https://github.com/rockset/rocksdb-cloud/blob/11877b32890419c92731fde5aa9213a47689782c/cloud/cloud_storage_provider.cc#L220

Reply:
Got it. So only if the manifest is persisted to s3, the writes are committed.